### DISCUSSION TITLE
Trane Z-Wave Programmable Thermostat (T400)

### DISCUSSION THREAD

**Discussion by gnsbglora** [zwave-js/zwave-js-ui#3541] [Q&A] [answered]
I'm using Z-Wave JS UI standalone with HomeAssistant MQTT gateway and MQTT discovery and trying to integrate a Trane Z-Wave Programmable Thermostat (T400). The thermostat is discovered and partially usable. (The same Thermostat worked before in a Z-Wave connection to Homeseer so it seems to be functional.) 
In HomeAssistantI can use off, heat, cool and auto mode however the configuration seems to be a little off. The discovery reports the min/max temperatures as 5/40 degrees meant to be Celsius but used as Fahrenheit because of the line "temperature_unit": "F". HomeAssistant is set to Fahrenheit.
1.  cooling mode
The min max temperatures I can override through the climate.thermostat_climate: min_temp max_temp in the HA configuration but that only seems to work for cooling mode.
2. heating mode
 When I activate heating mode I the target temperature goes somewhere unpredictable 5 degrees
3. auto mode
there's only one set point. shouldn't  there be two set points for the target temperature one for cooling and one for heating.

4. Fan mode
The control for Fan On/Auto is missing all together however it seems to exist in the discovery in the line "action_template"

Is there something I can change to make this work?

MQTT topic
```
homeassistant/climate/Thermostat/climate/config 

  "min_temp": 5,
  "max_temp": 40,
  "temp_step": 0.5,
  "modes": [
    "off",
    "heat",
    "cool",
    "auto"
  ],
  "mode_state_template": "{{ {0: \"off\", 1: \"heat\", 2: \"cool\", 3: \"auto\"}[value_json.value] | default('off') }}",
  "current_temperature_topic": "zwave/Thermostat/49/0/Air_temperature",
  "current_temperature_template": "{{ value_json.value }}",
  "temperature_state_topic": "zwave/Thermostat/67/0/setpoint/2",
  "temperature_state_template": "{{ value_json.value }}",
  "temperature_command_topic": "zwave/Thermostat/67/0/setpoint/2/set",
  "mode_state_topic": "zwave/Thermostat/64/0/mode",
  "mode_command_topic": "zwave/Thermostat/64/0/mode/set",
  "action_topic": "zwave/Thermostat/66/0/state",
  "action_template": "{{ {0: \"idle\", 1: \"heating\", 2: \"cooling\", 3: \"fan\", 4: \"idle\", 5: \"idle\", 6: \"fan\", 7: \"heating\", 8: \"heating\", 9: \"cooling\", 10: \"heating\", 11: \"heating\"}[value_json.value] | default('idle') }}",
  "temperature_unit": "F",
  "precision": 0.1,
  "device": {
    "identifiers": [
      "zwavejs2mqtt_0xxxxxxxxx_node30"
    ],
    "manufacturer": "Trane Corporation",
    "model": "Trane Z-Wave Programmable Thermostat (T400)",
    "name": "Thermostat",
    "sw_version": "20.9"
  },
  "availability": [
    {
      "payload_available": "true",
      "payload_not_available": "false",
      "topic": "zwave/Thermostat/status",
      "value_template": "{{'true' if value_json.value else 'false'}}"
    },
    {
      "topic": "zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/status",
      "value_template": "{{'online' if value_json.value else 'offline'}}"
    },
    {
      "payload_available": "true",
      "payload_not_available": "false",
      "topic": "zwave/driver/status"
    }
  ],
  "availability_mode": "all",
  "name": "Thermostat_climate",
  "unique_id": "zwavejs2mqtt_0xxxxxxxxx_Node30_climate"
}

```

**Comment by robertsLando** [comment]
If you're using HA, you should preferably use the Z-Wave integration with the Z-Wave JS UI addon and not MQTT. Integration has much better support of devices compared to the MQTT discovery solution

**Comment by gnsbglora** [comment]
In my Case HA is running as a virtual machine on a proxmox VE server and Z-Wave JS UI is installed trough Node.js on a diskless PXE-booted and poe-powered Raspberry PI. The proxmox server lives in a server rack too far away and the plan was to add another raspberry pi with Z-Wave JS UI for a second building in the future. Everything else seems to work fine except for this issue with the thermostat.

Is there a way to influence or override the discovery to add the missing fields?

**Reply by robertsLando** [reply]
> The proxmox server lives in a server rack too far away and the plan was to add another raspberry pi with Z-Wave JS UI for a second building in the future

I don't understand the problem with this, you just have to configure the Home assistant zwave integration to point to the server address that runs zui and you are done.

About your issue, problem is I don't use home assistant and I neighter have a thermostat to test discovery. Discovery termostats using MQTT is just a mess because every time the mode changes I need to rediscover them in order to fix the setpoint topics and so on... Apart from this there are also other things that make mqtt discovery really hard to maintain and debug, that's why I generally suggest to use home assistant integration where possible as that is the official (and feature rich) way

**Reply by gnsbglora** [reply]
It seems I didn’t understand how the pieces are supposed to work together. I didn’t install a z-wave integration on HA since it doesn’t have a z-wave stick/interface attached to it. It was my understanding if I want a detached z-wave satellite it would have to be 

HA <-> MQTT <-> Z-Wave JS UI 

I saw that there is also the option for the web socket interface but the HA documentation seems to be a little vague about the information flow. I tried to turn that on first, but nothing showed up on the HA side. Once I enabled the MQTT discovery the z-wave devices became visible on the HA side.

> you just have to configure the Home assistant zwave integration to point to the server address that runs zui and you are done

I’m assuming according to your comment I would have to install a z-wave integration on HA and point it to Z-Wave JS UI instead of using a local z-wave stick/interface. Where would I have to configure that server address in the HA zwave integration?

**Reply by robertsLando** [reply, marked as answer]
The stick must be connected to the instance running zwave-js-ui so if you have a server running hass and an rpi with zui it would be:

**Server A**...............................**Server B (RPI)**
Home Assistant ................... Zwave-JS UI
Zwave-JS Integration  <------> Websocket ( default port 3000)

Essentially when you enable `WS Server` option under `Home Assistant` section in UI Settings ZUI will open a websocket server that allows home assistant to communicate with the stick.

@kpine could give you more details about how to configure the integration propely

**Reply by gnsbglora** [reply]
Thank you very much! I will give that a try.

**Reply by gnsbglora** [reply]
It's working now. My mistake on the web socket settings was to put the HA server name into the zwave-js-ui -> settings -> home assistant -> ws server -> server host field. I assumed HA to be the web socket server and zwave-js-ui the client. The thermostat seems to be ok now.
Thank you for your help! Now I just have to clean out all the mqtt-doubles.
